extmod/uasyncio/stream.py

```diff
@@ -0,0 +1,141 @@
+# MicroPython uasyncio module
+# MIT license; Copyright (c) 2019-2020 Damien P. George
+
+from . import core
+
+
+class Stream:
+    def __init__(self, s, e={}):
+        self.s = s
+        self.e = e
+        self.out_buf = b""
+
+    def get_extra_info(self, v):
+        return self.e[v]
+
+    async def __aenter__(self):
+        return self
+
+    async def __aexit__(self, exc_type, exc, tb):
+        await self.close()
+
+    def close(self):
+        pass
+
+    async def wait_closed(self):
+        # TODO yield?
+        self.s.close()
+
+    async def read(self, n):
+        yield core._io_queue.queue_read(self.s)
+        return self.s.read(n)
+
+    async def readline(self):
+        l = b""
+        while True:
+            yield core._io_queue.queue_read(self.s)
+            l2 = self.s.readline()  # may do multiple reads but won't block
+            l += l2
+            if not l2 or l[-1] == 10:  # \n (check l in case l2 is str)
+                return l
+
+    def write(self, buf):
+        self.out_buf += buf
+
+    async def drain(self):
+        mv = memoryview(self.out_buf)
+        off = 0
+        while off < len(mv):
+            yield core._io_queue.queue_write(self.s)
+            ret = self.s.write(mv[off:])
+            if ret is not None:
+                off += ret
+        self.out_buf = b""
```

Comment:
On the esp32 `drain` fails to raise an error if the stream is closed. Test coro:

```
async def test_close_write():
    sr, sw = await asyncio.open_connection('192.168.0.14', 1883)
    print("connected")
    sw.close()
    await sw.wait_closed()
    print("closed")
    try:
        sw.write(b'hello')
        await sw.drain()
        assert True == False, "Error: drain on closed socket didn't raise"
    except OSError as e:
        print("Got OSError:", e)
```

Outputs "connected" and "closed" then hangs. Under CPython raises a ConnectionResetError as expected.

Reply:
ok, this will need to be fixed (probably separately to this PR)

Reply:
I've now added a test for this, `tests/multi_net/uasyncio_tcp_close_write.py`.  It passes (using unix MicroPython) without any changes to uasyncio, so I'm not sure why you were seeing this problem....

Reply:
The following fails for me, using an esp32 on dev/ttyUSB1:
```
> ./run-multitests.py -i exec:../ports/unix/micropython -s -i pyb:/dev/ttyUSB1 multi_net/uasyncio_tcp_close_write.py
multi_net/uasyncio_tcp_close_write.py on micropython|ttyUSB1:
### TEST ###
--- instance0 ---
server running
Traceback (most recent call last):
  File "<stdin>", line 91, in <module>
  File "<stdin>", line 52, in instance0
  File "../extmod/uasyncio/core.py", line 209, in run
  File "../extmod/uasyncio/core.py", line 189, in run_until_complete
  File "../extmod/uasyncio/core.py", line 178, in run_until_complete
  File "<stdin>", line 34, in tcp_server
  File "../extmod/uasyncio/funcs.py", line 28, in wait_for
TimeoutError:
--- instance1 ---
close
write
TIMEOUT
### TRUTH ###
--- instance0 ---
server running
echo: b''
close
done
--- instance1 ---
close
write
OSError
FAIL

1 tests performed
0 tests passed
1 tests failed: multi_net/uasyncio_tcp_close_write.py
```
I'll look some more into this.
NB: to figure out how to run this test I had to go to tests/run-multitests.py, go to the blame, click on the commit for the file, then read the commit info. :-(

Reply:
Here's another example:
```
> cat test_close_write.py
import uasyncio as asyncio
server = '192.168.0.2'

async def test_close_write():
    sr, sw = await asyncio.open_connection(server, 8899)
    print("connected")
    #await asyncio.sleep_ms(200)
    sw.write("Hi there!\n")
    await sw.drain()
    print("wrote")
    #await asyncio.sleep_ms(200)
    sw.close()
    await sw.wait_closed()
    print("closed")
    try:
        sw.write(b"hello")
        await sw.drain()
        assert True == False, "Error: drain on closed socket didn't raise"
    except OSError as e:
        print("Got OSError:", e)

loop = asyncio.get_event_loop()
loop.run_until_complete(test_close_write())
print("done")
```
On the "server" machine I run:
```
> nc -l 8899
Hi there!
>
```
and the transcript on the esp32:
```
> ./pyb test_close_write.py
connected
wrote
closed
^C⏎
```
I also ran this on micropython/linux by adding something like this at the top of the test:
```
import sys
sys.path.append('/home/src/esp32/micropython/extmod')
```
The transcript is:
```
> ../../micropython/ports/unix/micropython test_close_write.py
connected
wrote
closed
Got OSError: [Errno 9] EBADF
done
```

Reply:
Looking at the code the unix implementation of poll/select/streams is completely different from esp32 and much simpler 'cause all streams are file descriptors...

I find https://github.com/micropython/micropython/blob/master/ports/esp32/modsocket.c#L714-L730 suspect. For one, it doesn't seem to support half-close? But for the issue at hand it sets socket->fd to -1 and then what is the next MP_STREAM_POLL request supposed to return? Is there a `if (socket->fd < 0) return MP_STREAM_POLL_HUP` missing?

Reply:
Ok, thanks, I can now reproduce this problem on esp32.  It works on a PYBD using modlwip so it's an esp32 vs the rest issue.  Actually, this behaviour is tested by `tests/extmod/uselect_poll_basic.py` which currently fails on esp32 so would need to be fixed separately.

> Add-on question: why does the esp32 not use modlwip?

Because lwip on the esp32 is a modified version of lwip to support multi-core and it's not possible to directly call into the lwip "low level" API, like `modlwip.c` does.